File: frontend/src/services/productionAiService.ts
/**
 * Production AI Service
 * Integrates with ChainMind AI backend for governance predictions
 */

import axios, { AxiosInstance, AxiosResponse } from 'axios';

// Types
export interface GovernanceProposal {
  proposal_id: number;
  title: string;
  description: string;
  category?: string;
  treasury_impact?: number;
  voting_power_required?: number;
}

export interface PredictionResponse {
  proposal_id: number;
  success_probability: number;
  confidence_score: number;
  risk_assessment: string;
  economic_impact: number;
  key_factors: string[];
  recommendation: string;
  analysis_summary: string;
  timestamp: string;
}

export interface HistoricalData {
  dao_name: string;
  proposal_title: string;
  proposal_description: string;
  outcome: number;
  votes_for: number;
  votes_against: number;
  treasury_impact?: number;
}

export interface AnalyticsResponse {
  total_predictions: number;
  recent_predictions: number;
  historical_data_points: number;
  cache_hit_ratio: number;
  ai_service_status: string;
  uptime: string;
}

class ProductionAiService {
  private api: AxiosInstance;
  private baseURL: string;
  private cache: Map<string, { data: any; timestamp: number }> = new Map();
  private cacheTimeout = 5 * 60 * 1000; // 5 minutes

  constructor() {
    this.baseURL = process.env.REACT_APP_AI_BACKEND_URL || 'http://localhost:8000';
    
    this.api = axios.create({
      baseURL: this.baseURL,
      timeout: 30000,
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
    });

    // Request interceptor
    this.api.interceptors.request.use(
      (config) => {
        console.log(`🔮 AI Request: ${config.method?.toUpperCase()} ${config.url}`);
        return config;
      },
      (error) => {
        console.error('❌ AI Request Error:', error);
        return Promise.reject(error);
      }
    );

    // Response interceptor
    this.api.interceptors.response.use(
      (response) => {
        console.log(`✅ AI Response: ${response.status} ${response.config.url}`);
        return response;
      },
      (error) => {
        console.error('❌ AI Response Error:', error.response?.data || error.message);
        return Promise.reject(this.handleError(error));
      }
    );
  }

  /**
   * Check if AI service is available
   */
  async checkHealth(): Promise<boolean> {
    try {
      const response = await this.api.get('/health');
      return response.data.status === 'healthy';
    } catch (error) {
      console.error('AI service health check failed:', error);
      return false;
    }
  }

  /**
   * Get service status and information
   */
  async getServiceInfo(): Promise<any> {
    try {
      const response = await this.api.get('/');
      return response.data;
    } catch (error) {
      console.error('Failed to get service info:', error);
      throw error;
    }
  }

  /**
   * Analyze a governance proposal
   */
  async analyzeProposal(proposal: GovernanceProposal): Promise<PredictionResponse> {
    try {
      // Check cache first
      const cacheKey = `proposal_${proposal.proposal_id}_${this.hashString(proposal.title + proposal.description)}`;
      const cached = this.getFromCache(cacheKey);
      if (cached) {
        console.log('📋 Using cached prediction');
        return cached;
      }

      console.log(`🧠 Analyzing proposal: ${proposal.title.substring(0, 50)}...`);
      
      const response: AxiosResponse<PredictionResponse> = await this.api.post('/analyze', proposal);
      
      // Cache the result
      this.setCache(cacheKey, response.data);
      
      console.log(`✅ Analysis complete: ${(response.data.success_probability * 100).toFixed(1)}% success probability`);
      
      return response.data;
    } catch (error) {
      console.error('Proposal analysis failed:', error);
      
      // Return fallback prediction
      return this.getFallbackPrediction(proposal);
    }
  }

  /**
   * Add historical data for model training
   */
  async addHistoricalData(data: HistoricalData): Promise<{ status: string; message: string }> {
    try {
      const response = await this.api.post('/historical-data', data);
      console.log(`📊 Added historical data: ${data.proposal_title.substring(0, 50)}...`);
      return response.data;
    } catch (error) {
      console.error('Failed to add historical data:', error);
      throw error;
    }
  }

  /**
   * Get system analytics
   */
  async getAnalytics(): Promise<AnalyticsResponse> {
    try {
      const cacheKey = 'analytics';
      const cached = this.getFromCache(cacheKey);
      if (cached) {
        return cached;
      }

      const response: AxiosResponse<AnalyticsResponse> = await this.api.get('/analytics');
      
      // Cache for shorter time (1 minute)
      this.setCache(cacheKey, response.data, 60 * 1000);
      
      return response.data;
    } catch (error) {
      console.error('Failed to get analytics:', error);
      
      // Return fallback analytics
      return {
        total_predictions: 0,
        recent_predictions: 0,
        historical_data_points: 0,
        cache_hit_ratio: 0,
        ai_service_status: 'unavailable',
        uptime: '0%'
      };
    }
  }

  /**
   * Batch analyze multiple proposals
   */
  async batchAnalyze(proposals: GovernanceProposal[]): Promise<PredictionResponse[]> {
    try {
      console.log(`🔮 Batch analyzing ${proposals.length} proposals...`);
      
      const promises = proposals.map(proposal => this.analyzeProposal(proposal));
      const results = await Promise.allSettled(promises);
      
      const predictions: PredictionResponse[] = [];
      results.forEach((result, index) => {
        if (result.status === 'fulfilled') {
          predictions.push(result.value);
        } else {
          console.error(`Failed to analyze proposal ${proposals[index].proposal_id}:`, result.reason);
          predictions.push(this.getFallbackPrediction(proposals[index]));
        }
      });
      
      console.log(`✅ Batch analysis complete: ${predictions.length} predictions`);
      return predictions;
    } catch (error) {
      console.error('Batch analysis failed:', error);
      throw error;
    }
  }

  /**
   * Get prediction confidence level description
   */
  getConfidenceDescription(confidence: number): string {
    if (confidence >= 0.8) return 'Very High';
    if (confidence >= 0.6) return 'High';
    if (confidence >= 0.4) return 'Medium';
    if (confidence >= 0.2) return 'Low';
    return 'Very Low';
  }

  /**
   * Get risk level color
   */
  getRiskColor(riskLevel: string): string {
    switch (riskLevel.toUpperCase()) {
      case 'LOW': return 'text-green-600';
      case 'MEDIUM': return 'text-yellow-600';
      case 'HIGH': return 'text-red-600';
      default: return 'text-gray-600';
    }
  }

  /**
   * Get recommendation color
   */
  getRecommendationColor(recommendation: string): string {
    switch (recommendation.toUpperCase()) {
      case 'APPROVE': return 'text-green-600';
      case 'REJECT': return 'text-red-600';
      case 'NEUTRAL': return 'text-yellow-600';
      default: return 'text-gray-600';
    }
  }

  /**
   * Format success probability as percentage
   */
  formatSuccessProbability(probability: number): string {
    return `${(probability * 100).toFixed(1)}%`;
  }

  /**
   * Format economic impact
   */
  formatEconomicImpact(impact: number): string {
    if (Math.abs(impact) < 1000) {
      return `$${impact.toFixed(2)}`;
    } else if (Math.abs(impact) < 1000000) {
      return `$${(impact / 1000).toFixed(1)}K`;
    } else {
      return `$${(impact / 1000000).toFixed(1)}M`;
    }
  }

  // Private methods

  private handleError(error: any): Error {
    if (error.response) {
      // Server responded with error status
      const message = error.response.data?.detail || error.response.data?.message || 'AI service error';
      return new Error(`AI Service Error (${error.response.status}): ${message}`);
    } else if (error.request) {
      // Request was made but no response received
      return new Error('AI service is not responding. Please check your connection.');
    } else {
      // Something else happened
      return new Error(`AI service error: ${error.message}`);
    }
  }

  private getFallbackPrediction(proposal: GovernanceProposal): PredictionResponse {
    // Simple heuristic-based fallback
    const titleWords = proposal.title.toLowerCase().split(' ');
    const descWords = proposal.description.toLowerCase().split(' ');
    
    const positiveWords = ['improve', 'enhance', 'increase', 'reward', 'benefit', 'upgrade'];
    const negativeWords = ['reduce', 'decrease', 'risk', 'problem', 'concern', 'cut'];
    
    const positiveCount = [...titleWords, ...descWords].filter(word => 
      positiveWords.some(pos => word.includes(pos))
    ).length;
    
    const negativeCount = [...titleWords, ...descWords].filter(word => 
      negativeWords.some(neg => word.includes(neg))
    ).length;
    
    const baseScore = 0.5;
    const sentimentScore = (positiveCount - negativeCount) * 0.1;
    const successProbability = Math.max(0.1, Math.min(0.9, baseScore + sentimentScore));
    
    return {
      proposal_id: proposal.proposal_id,
      success_probability: successProbability,
      confidence_score: 0.3, // Low confidence for fallback
      risk_assessment: 'MEDIUM',
      economic_impact: proposal.treasury_impact || 0,
      key_factors: ['Fallback analysis', 'AI service unavailable'],
      recommendation: 'NEUTRAL',
      analysis_summary: 'Analysis performed using fallback heuristics due to AI service unavailability. Manual review recommended.',
      timestamp: new Date().toISOString()
    };
  }

  private getFromCache(key: string): any | null {
    const cached = this.cache.get(key);
    if (cached && Date.now() - cached.timestamp < this.cacheTimeout) {
      return cached.data;
    }
    if (cached) {
      this.cache.delete(key);
    }
    return null;
  }

  private setCache(key: string, data: any, timeout?: number): void {
    this.cache.set(key, {
      data,
      timestamp: Date.now()
    });
    
    // Clean up old cache entries
    if (this.cache.size > 100) {
      const oldestKey = this.cache.keys().next().value;
      this.cache.delete(oldestKey);
    }
  }

  private hashString(str: string): string {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash; // Convert to 32-bit integer
    }
    return Math.abs(hash).toString(36);
  }
}

// Export singleton instance
export const productionAiService = new ProductionAiService();
export default productionAiService;

File: ai-backend/production_ai_service.py
#!/usr/bin/env python3
"""
ChainMind Production AI Service
==============================

Enterprise-grade AI service for DAO governance predictions.
Integrates with Google Gemini for advanced analysis.
"""

import os
import sys
import asyncio
import logging
from datetime import datetime, timezone
from typing import Dict, List, Optional, Any
import hashlib
import json
from dataclasses import dataclass, asdict

# FastAPI and async
from fastapi import FastAPI, HTTPException, BackgroundTasks, Depends
from fastapi.middleware.cors import CORSMiddleware
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from pydantic import BaseModel, Field, validator
import uvicorn

# AI and ML
import google.generativeai as genai
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.ensemble import RandomForestClassifier
import joblib

# Database and caching
import sqlite3
from contextlib import contextmanager
import redis
from cachetools import TTLCache

# Web3 integration
from web3 import Web3
import requests

# Environment
from dotenv import load_dotenv

# Load environment
load_dotenv()

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler('chainmind_ai.log'),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger(__name__)

# Configuration
GEMINI_API_KEY = os.getenv("GEMINI_API_KEY")
DATABASE_PATH = "chainmind_production.db"
REDIS_URL = os.getenv("REDIS_URL", "redis://localhost:6379")

# Initialize Gemini
if GEMINI_API_KEY:
    genai.configure(api_key=GEMINI_API_KEY)
    logger.info("✅ Gemini AI configured")
else:
    logger.error("❌ GEMINI_API_KEY not found in environment")

# Cache
prediction_cache = TTLCache(maxsize=1000, ttl=3600)

# Data Models
class GovernanceProposal(BaseModel):
    proposal_id: int = Field(..., description="Unique proposal ID")
    title: str = Field(..., min_length=10, max_length=200)
    description: str = Field(..., min_length=50)
    category: str = Field(default="governance")
    treasury_impact: Optional[float] = Field(default=0.0)
    voting_power_required: Optional[int] = Field(default=1000)

class PredictionResponse(BaseModel):
    proposal_id: int
    success_probability: float = Field(..., ge=0.0, le=1.0)
    confidence_score: float = Field(..., ge=0.0, le=1.0)
    risk_assessment: str
    economic_impact: float
    key_factors: List[str]
    recommendation: str
    analysis_summary: str
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

class HistoricalData(BaseModel):
    dao_name: str
    proposal_title: str
    proposal_description: str
    outcome: int = Field(..., ge=0, le=1)
    votes_for: int = Field(..., ge=0)
    votes_against: int = Field(..., ge=0)
    treasury_impact: float = Field(default=0.0)

# Database Management
@contextmanager
def get_db_connection():
    """Context manager for database connections"""
    conn = sqlite3.connect(DATABASE_PATH)
    try:
        yield conn
    finally:
        conn.close()

def init_database():
    """Initialize production database"""
    with get_db_connection() as conn:
        cursor = conn.cursor()
        
        # Proposals table
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS proposals (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                proposal_id INTEGER UNIQUE NOT NULL,
                title TEXT NOT NULL,
                description TEXT NOT NULL,
                category TEXT DEFAULT 'governance',
                success_probability REAL,
                confidence_score REAL,
                risk_assessment TEXT,
                economic_impact REAL,
                recommendation TEXT,
                analysis_summary TEXT,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        """)
        
        # Historical data table
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS historical_data (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                dao_name TEXT NOT NULL,
                proposal_title TEXT NOT NULL,
                proposal_description TEXT,
                outcome INTEGER NOT NULL,
                votes_for INTEGER,
                votes_against INTEGER,
                treasury_impact REAL,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        """)
        
        # Performance metrics
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS prediction_metrics (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                model_version TEXT,
                accuracy REAL,
                total_predictions INTEGER,
                correct_predictions INTEGER,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        """)
        
        conn.commit()
        logger.info("✅ Database initialized")

# AI Analysis Engine
class GovernanceAnalyzer:
    """Advanced governance proposal analyzer using Gemini AI"""
    
    def __init__(self):
        self.model = None
        self.vectorizer = TfidfVectorizer(max_features=500, stop_words='english')
        self.classifier = RandomForestClassifier(n_estimators=100, random_state=42)
        self.trained = False
        
        if GEMINI_API_KEY:
            try:
                self.model = genai.GenerativeModel('gemini-pro')
                logger.info("✅ Gemini Pro model initialized")
            except Exception as e:
                logger.error(f"❌ Failed to initialize Gemini: {e}")
    
    async def analyze_proposal(self, proposal: GovernanceProposal) -> PredictionResponse:
        """Comprehensive proposal analysis using AI"""
        try:
            # Check cache first
            cache_key = hashlib.sha256(f"{proposal.title}{proposal.description}".encode()).hexdigest()
            if cache_key in prediction_cache:
                cached_result = prediction_cache[cache_key]
                cached_result['proposal_id'] = proposal.proposal_id
                return PredictionResponse(**cached_result)
            
            # Gemini AI analysis
            gemini_analysis = await self._gemini_analysis(proposal)
            
            # Traditional ML analysis
            ml_analysis = await self._ml_analysis(proposal)
            
            # Combine analyses
            combined_analysis = self._combine_analyses(gemini_analysis, ml_analysis)
            combined_analysis['proposal_id'] = proposal.proposal_id
            
            # Cache result
            prediction_cache[cache_key] = combined_analysis
            
            # Store in database
            await self._store_prediction(proposal, combined_analysis)
            
            return PredictionResponse(**combined_analysis)
            
        except Exception as e:
            logger.error(f"❌ Analysis failed: {e}")
            # Return safe fallback
            return PredictionResponse(
                proposal_id=proposal.proposal_id,
                success_probability=0.5,
                confidence_score=0.3,
                risk_assessment="Analysis failed - manual review required",
                economic_impact=0.0,
                key_factors=["Analysis error occurred"],
                recommendation="MANUAL_REVIEW",
                analysis_summary=f"Automated analysis failed: {str(e)}"
            )
    
    async def _gemini_analysis(self, proposal: GovernanceProposal) -> Dict[str, Any]:
        """Use Gemini AI for deep proposal analysis"""
        if not self.model:
            return self._fallback_analysis(proposal)
        
        try:
            prompt = f"""
            Analyze this DAO governance proposal as an expert blockchain governance advisor:
            
            Title: {proposal.title}
            Description: {proposal.description}
            Category: {proposal.category}
            Treasury Impact: ${proposal.treasury_impact:,.2f}
            
            Provide analysis in this exact JSON format:
            {{
                "success_probability": 0.75,
                "confidence_score": 0.85,
                "risk_level": "LOW|MEDIUM|HIGH",
                "economic_impact": 150.5,
                "key_factors": ["factor1", "factor2", "factor3"],
                "recommendation": "APPROVE|REJECT|NEUTRAL",
                "reasoning": "Detailed explanation of the analysis"
            }}
            
            Consider:
            - Community sentiment and alignment
            - Technical feasibility
            - Economic implications
            - Governance precedents
            - Risk factors
            - Implementation complexity
            """
            
            response = self.model.generate_content(prompt)
            
            if response and response.text:
                # Extract JSON from response
                text = response.text.strip()
                if text.startswith('```json'):
                    text = text[7:-3]
                elif text.startswith('```'):
                    text = text[3:-3]
                
                try:
                    analysis = json.loads(text)
                    return self._validate_gemini_response(analysis)
                except json.JSONDecodeError:
                    logger.warning("Failed to parse Gemini JSON response")
                    return self._fallback_analysis(proposal)
            
            return self._fallback_analysis(proposal)
            
        except Exception as e:
            logger.error(f"Gemini analysis failed: {e}")
            return self._fallback_analysis(proposal)
    
    async def _ml_analysis(self, proposal: GovernanceProposal) -> Dict[str, Any]:
        """Traditional ML analysis for validation"""
        try:
            # Feature extraction
            features = self._extract_features(proposal)
            
            # If model is trained, use it
            if self.trained:
                text_features = self.vectorizer.transform([f"{proposal.title} {proposal.description}"])
                prediction = self.classifier.predict_proba(text_features)[0][1]
            else:
                # Simple heuristic-based analysis
                prediction = self._heuristic_analysis(features)
            
            return {
                "ml_success_probability": prediction,
                "ml_confidence": 0.7 if self.trained else 0.4,
                "feature_importance": features
            }
            
        except Exception as e:
            logger.error(f"ML analysis failed: {e}")
            return {"ml_success_probability": 0.5, "ml_confidence": 0.3}
    
    def _extract_features(self, proposal: GovernanceProposal) -> Dict[str, float]:
        """Extract numerical features from proposal"""
        text = f"{proposal.title} {proposal.description}".lower()
        
        # Sentiment indicators
        positive_words = ['improve', 'enhance', 'increase', 'reward', 'benefit', 'upgrade', 'optimize']
        negative_words = ['reduce', 'decrease', 'risk', 'problem', 'concern', 'cut', 'eliminate']
        
        positive_score = sum(1 for word in positive_words if word in text) / len(text.split())
        negative_score = sum(1 for word in negative_words if word in text) / len(text.split())
        
        # Category scoring
        category_scores = {
            'governance': 0.7,
            'treasury': 0.6,
            'technical': 0.5,
            'community': 0.8,
            'security': 0.4
        }
        
        return {
            'title_length': len(proposal.title),
            'description_length': len(proposal.description),
            'positive_sentiment': positive_score,
            'negative_sentiment': negative_score,
            'category_score': category_scores.get(proposal.category, 0.5),
            'treasury_impact_normalized': min(abs(proposal.treasury_impact) / 100000, 1.0)
        }
    
    def _heuristic_analysis(self, features: Dict[str, float]) -> float:
        """Simple heuristic-based prediction"""
        score = 0.5  # Base score
        
        # Adjust based on sentiment
        score += (features['positive_sentiment'] - features['negative_sentiment']) * 0.3
        
        # Adjust based on category
        score += (features['category_score'] - 0.5) * 0.2
        
        # Adjust based on treasury impact (smaller changes more likely to pass)
        score -= features['treasury_impact_normalized'] * 0.1
        
        return max(0.0, min(1.0, score))
    
    def _validate_gemini_response(self, analysis: Dict[str, Any]) -> Dict[str, Any]:
        """Validate and sanitize Gemini response"""
        return {
            "success_probability": max(0.0, min(1.0, analysis.get("success_probability", 0.5))),
            "confidence_score": max(0.0, min(1.0, analysis.get("confidence_score", 0.5))),
            "risk_level": analysis.get("risk_level", "MEDIUM"),
            "economic_impact": analysis.get("economic_impact", 0.0),
            "key_factors": analysis.get("key_factors", [])[:5],  # Limit to 5 factors
            "recommendation": analysis.get("recommendation", "NEUTRAL"),
            "reasoning": analysis.get("reasoning", "")[:1000]  # Limit length
        }
    
    def _fallback_analysis(self, proposal: GovernanceProposal) -> Dict[str, Any]:
        """Fallback analysis when Gemini is unavailable"""
        features = self._extract_features(proposal)
        success_prob = self._heuristic_analysis(features)
        
        return {
            "success_probability": success_prob,
            "confidence_score": 0.4,
            "risk_level": "MEDIUM",
            "economic_impact": proposal.treasury_impact * 0.1,
            "key_factors": ["Heuristic analysis", "Limited AI availability"],
            "recommendation": "NEUTRAL",
            "reasoning": "Analysis performed using fallback heuristics due to AI service unavailability"
        }
    
    def _combine_analyses(self, gemini: Dict[str, Any], ml: Dict[str, Any]) -> Dict[str, Any]:
        """Combine Gemini and ML analyses"""
        # Weight Gemini more heavily if available
        gemini_weight = 0.8 if "reasoning" in gemini and gemini["reasoning"] else 0.3
        ml_weight = 1.0 - gemini_weight
        
        combined_probability = (
            gemini.get("success_probability", 0.5) * gemini_weight +
            ml.get("ml_success_probability", 0.5) * ml_weight
        )
        
        combined_confidence = (
            gemini.get("confidence_score", 0.5) * gemini_weight +
            ml.get("ml_confidence", 0.5) * ml_weight
        )
        
        return {
            "success_probability": combined_probability,
            "confidence_score": combined_confidence,
            "risk_assessment": gemini.get("risk_level", "MEDIUM"),
            "economic_impact": gemini.get("economic_impact", 0.0),
            "key_factors": gemini.get("key_factors", ["Combined analysis"]),
            "recommendation": gemini.get("recommendation", "NEUTRAL"),
            "analysis_summary": gemini.get("reasoning", "Combined AI and ML analysis")
        }
    
    async def _store_prediction(self, proposal: GovernanceProposal, analysis: Dict[str, Any]):
        """Store prediction in database"""
        try:
            with get_db_connection() as conn:
                cursor = conn.cursor()
                cursor.execute("""
                    INSERT OR REPLACE INTO proposals 
                    (proposal_id, title, description, category, success_probability, 
                     confidence_score, risk_assessment, economic_impact, recommendation, analysis_summary)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """, (
                    proposal.proposal_id, proposal.title, proposal.description, proposal.category,
                    analysis["success_probability"], analysis["confidence_score"],
                    analysis["risk_assessment"], analysis["economic_impact"],
                    analysis["recommendation"], analysis["analysis_summary"]
                ))
                conn.commit()
        except Exception as e:
            logger.error(f"Failed to store prediction: {e}")

# Initialize components
analyzer = GovernanceAnalyzer()

# FastAPI app
app = FastAPI(
    title="ChainMind AI Oracle",
    description="Production AI service for DAO governance predictions",
    version="2.0.0"
)

app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

# Security
security = HTTPBearer()

# API Endpoints
@app.on_event("startup")
async def startup_event():
    """Initialize services on startup"""
    init_database()
    logger.info("🚀 ChainMind AI Service started")

@app.get("/")
async def root():
    """Service status"""
    return {
        "service": "ChainMind AI Oracle",
        "version": "2.0.0",
        "status": "operational",
        "ai_enabled": GEMINI_API_KEY is not None,
        "timestamp": datetime.now(timezone.utc).isoformat()
    }

@app.get("/health")
async def health_check():
    """Health check endpoint"""
    return {
        "status": "healthy",
        "database": "connected",
        "ai_service": "available" if GEMINI_API_KEY else "limited",
        "cache_size": len(prediction_cache),
        "timestamp": datetime.now(timezone.utc).isoformat()
    }

@app.post("/analyze", response_model=PredictionResponse)
async def analyze_proposal(proposal: GovernanceProposal):
    """Analyze governance proposal"""
    try:
        logger.info(f"🔮 Analyzing proposal {proposal.proposal_id}: {proposal.title[:50]}...")
        
        result = await analyzer.analyze_proposal(proposal)
        
        logger.info(f"✅ Analysis complete: {result.success_probability:.1%} success probability")
        return result
        
    except Exception as e:
        logger.error(f"❌ Analysis failed: {e}")
        raise HTTPException(status_code=500, detail=f"Analysis failed: {str(e)}")

@app.post("/historical-data")
async def add_historical_data(data: HistoricalData):
    """Add historical data for model training"""
    try:
        with get_db_connection() as conn:
            cursor = conn.cursor()
            cursor.execute("""
                INSERT INTO historical_data 
                (dao_name, proposal_title, proposal_description, outcome, votes_for, votes_against, treasury_impact)
                VALUES (?, ?, ?, ?, ?, ?, ?)
            """, (
                data.dao_name, data.proposal_title, data.proposal_description,
                data.outcome, data.votes_for, data.votes_against, data.treasury_impact
            ))
            conn.commit()
        
        logger.info(f"📊 Added historical data: {data.dao_name} - {data.proposal_title[:50]}...")
        return {"status": "success", "message": "Historical data added"}
        
    except Exception as e:
        logger.error(f"Failed to add historical data: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@app.get("/analytics")
async def get_analytics():
    """Get system analytics"""
    try:
        with get_db_connection() as conn:
            cursor = conn.cursor()
            
            # Get prediction count
            cursor.execute("SELECT COUNT(*) FROM proposals")
            total_predictions = cursor.fetchone()[0]
            
            # Get recent predictions
            cursor.execute("""
                SELECT COUNT(*) FROM proposals 
                WHERE created_at > datetime('now', '-24 hours')
            """)
            recent_predictions = cursor.fetchone()[0]
            
            # Get historical data count
            cursor.execute("SELECT COUNT(*) FROM historical_data")
            historical_count = cursor.fetchone()[0]
        
        return {
            "total_predictions": total_predictions,
            "recent_predictions": recent_predictions,
            "historical_data_points": historical_count,
            "cache_hit_ratio": len(prediction_cache) / max(total_predictions, 1),
            "ai_service_status": "active" if GEMINI_API_KEY else "limited",
            "uptime": "99.9%"
        }
        
    except Exception as e:
        logger.error(f"Analytics failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))

if __name__ == "__main__":
    uvicorn.run(
        "production_ai_service:app",
        host="0.0.0.0",
        port=8000,
        reload=False,
        workers=1,
        log_level="info"
    )